drivers/stepper/adi_tmc/adi_tmc5160_stepper_controller.c

```diff
@@ -0,0 +1,670 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Stefano Cottafavi
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_tmc5160
+
+#include <stdlib.h>
+
+#include <zephyr/drivers/stepper.h>
+#include <zephyr/drivers/stepper/stepper_trinamic.h>
+
+#include "adi_tmc_reg.h"
+#include "adi_tmc_spi.h"
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(tmc5160, CONFIG_STEPPER_LOG_LEVEL);
+
+struct tmc5160_data {
+	struct k_sem sem;
+};
+
+struct tmc5160_config {
+	const uint32_t gconf;
+	struct spi_dt_spec spi;
+	const uint32_t clock_frequency;
+};
+
+struct tmc5160_stepper_data {
+	struct k_work_delayable stallguard_dwork;
+	/* Work item to run the callback in a thread context. */
+#ifdef CONFIG_STEPPER_ADI_TMC5160_RAMPSTAT_POLL
+	struct k_work_delayable rampstat_callback_dwork;
+#endif
+	/* device pointer required to access config in k_work */
+	const struct device *stepper;
+	stepper_event_callback_t callback;
+	void *event_cb_user_data;
+};
+
+struct tmc5160_stepper_config {
+	const uint8_t index;
+	const uint16_t default_micro_step_res;
+	const int8_t sg_threshold;
+	const bool is_sg_enabled;
+	const uint32_t sg_velocity_check_interval_ms;
+	const uint32_t sg_threshold_velocity;
+	/* parent controller required for bus communication */
+	const struct device *controller;
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+	const struct tmc_ramp_generator_data default_ramp_config;
+#endif
+};
+
+static int tmc5160_write(const struct device *dev, const uint8_t reg_addr, const uint32_t reg_val)
+{
+	const struct tmc5160_config *config = dev->config;
+	struct tmc5160_data *data = dev->data;
```

Comment:
would it make sense to use the child device as argument to `tmc5160_write`, change these to `dev->controller->config` and `dev->controller->data` and just pass `dev` to this API? Make the code a bit more compact